spec/index.md

```diff
@@ -1289,6 +1293,18 @@ While an implementation will likely try to keep the interval between the value o
 
 :::
 
+#### On Low Wasm Memory {#on-low-wasm-memory}
+
+A canister can export a function with the name `canister_on_low_wasm_memory`, which is scheduled whenever the canister's wasm memory size in bytes grows from below a threshold `t` to >= `t`.
```

Comment:
@mraszyk WDYT? I think that more important information is how much memory is left. Furthermore, if we decide to check the replica's remaining memory, we cannot use the upper bound on the wasm memory size for that.

Reply:
Makes sense to me. Should the inequality be <= or < ?

Reply:
"<" makes more sense, since the default is 0, hence if we have a default value of the treshold the hook will never be invoked. I updated sugesstion.

Reply:
Discussion in the interface spec meeting:
- The semantics here are unclear, for instance in the case that the canister has a certain memory allocation and uses both heap and stable memory. What does "the canister's remaining wasm memory size" then refer to, exactly?
- The formulation seems to indicate a guarantee that actually cannot be provided (without an explicit memory allocation): Suppose some canister makes a big memory allocation, which brings some 10s of thousands of canisters on the same subnet over/under the threshold. We would have to call the hook on all of those 10s of thousands of canisters, but it's very unlikely that all of those would be able to satisfy their memory needs.

Thus, we are not convinced that this change of semantics is beneficial, given that the only way for a canister to guarantee that it can still allocate memory (from the shared resource of overall subnet memory) is via memory allocation, which however has ambiguous semantics as described above.

Reply:
Another complication here is that the canister may not be able to _pay for_ allocating additional memory due to the cycles prepayment when the subnet memory becomes scarce. Should that trigger the condition as well?

Reply:
I agree it is better to look only at the `canister's memory`, not the `subnet available memory`. In that case:

1. If the certain application uses both heap and stable memory then the `remaining wasm memory` will be:

min(`memory_allocation` - `used_stable_memory`, 4 GB) - `used_heap_memory`

2. If we look only at the canister's memory, the allocation of the other canister cannot trigger the hook on the current canister, hence that will not be a problem. Furthermore, by checking if the `remaining wasm memory` is below the `threshold` we will ensure enough `wasm memory` in case of usage of `memory allocation` (first part in the formula above), which cannot be done if the hook is checking if the `wasm memory` if above the `threshold`.

Reply:
The formula would likely have to include a few more terms that are accounted for in `memory_allocation`, such as the wasm module size or the size of the canister history. When that is done, it seems an acceptable API.

I guess in my opinion there is no clearly "better" API. The "old" one is a bit easier to understand in that it uses more basic values, but has the disadvantage of probably requiring the canister to change the threshold at least upon significant allocation of additional stable memory. This "new" variant may be a bit easier to use, but is bit more complicated to understand and would need to be adapted when new features (e.g. multiple wasm memories) are supported.

Reply:
In that case, I will go forward and update the specs to use “remaining memory drops to below a threshold”.